gadget/device_test.go

```diff
@@ -194,8 +219,158 @@ func (d *deviceSuite) TestDeviceFindNotFoundNotASymlink(c *C) {
+	c.Check(err, ErrorMatches, `device not found`)
+	c.Check(found, Equals, "")
+	c.Check(offs, Equals, gadget.Size(0))
+}
+
+func (d *deviceSuite) TestDeviceFindFallbackBadWritable(c *C) {
+	err := ioutil.WriteFile(filepath.Join(d.dir, "/dev/fakedevice0p1"), []byte(""), 0644)
+	c.Assert(err, IsNil)
+	err = os.Symlink("../../fakedevice0p1", filepath.Join(d.dir, "/dev/disk/by-label/writable"))
+	c.Assert(err, IsNil)
+
+	ps := &gadget.PositionedStructure{
+		VolumeStructure: &gadget.VolumeStructure{
+			Type: "bare",
+		},
+		StartOffset: 123,
+	}
+
+	found, offs, err := gadget.FindDeviceForStructureWithFallback(ps)
+	c.Check(err, ErrorMatches, `unexpected number of matches \(0\) for /sys/block/\*/fakedevice0p1`)
+	c.Check(found, Equals, "")
+	c.Check(offs, Equals, gadget.Size(0))
+
+	err = os.MkdirAll(filepath.Join(d.dir, "/sys/block/fakedevice0/fakedevice0p1"), 0755)
+	c.Assert(err, IsNil)
+
+	found, offs, err = gadget.FindDeviceForStructureWithFallback(ps)
+	c.Check(err, ErrorMatches, `device .*/dev/fakedevice0 does not exist`)
 	c.Check(found, Equals, "")
+	c.Check(offs, Equals, gadget.Size(0))
+}
+
+func (d *deviceSuite) TestDeviceFindFallbackHappyWritable(c *C) {
+	d.setUpWritableFallback(c)
+
+	psJustBare := &gadget.PositionedStructure{
+		VolumeStructure: &gadget.VolumeStructure{
+			Type: "bare",
+		},
+		StartOffset: 123,
+	}
+	psBareWithName := &gadget.PositionedStructure{
+		VolumeStructure: &gadget.VolumeStructure{
+			Type: "bare",
+			Name: "foo",
+		},
+		StartOffset: 123,
+	}
+	psNoName := &gadget.PositionedStructure{
+		VolumeStructure: &gadget.VolumeStructure{},
+		StartOffset:     123,
+	}
+
+	for _, ps := range []*gadget.PositionedStructure{psJustBare, psBareWithName, psNoName} {
+		found, offs, err := gadget.FindDeviceForStructureWithFallback(ps)
+		c.Check(err, IsNil)
+		c.Check(found, Equals, filepath.Join(d.dir, "/dev/fakedevice0"))
+		c.Check(offs, Equals, gadget.Size(123))
+	}
+}
+
+func (d *deviceSuite) TestDeviceFindFallbackNotForNamedWritable(c *C) {
+	d.setUpWritableFallback(c)
+
+	// should not hit the fallback path
+	psNamed := &gadget.PositionedStructure{
+		VolumeStructure: &gadget.VolumeStructure{
+			Name: "foo",
+		},
+		StartOffset: 123,
+	}
+	found, offs, err := gadget.FindDeviceForStructureWithFallback(psNamed)
+	c.Check(err, Equals, gadget.ErrDeviceNotFound)
+	c.Check(found, Equals, "")
+	c.Check(offs, Equals, gadget.Size(0))
+}
+
+func (d *deviceSuite) TestDeviceFindFallbackNotForFilesystem(c *C) {
+	d.setUpWritableFallback(c)
+
+	psFs := &gadget.PositionedStructure{
+		VolumeStructure: &gadget.VolumeStructure{
+			Label:      "foo",
+			Filesystem: "ext4",
+		},
+		StartOffset: 123,
+	}
+	found, offs, err := gadget.FindDeviceForStructureWithFallback(psFs)
+	c.Check(err, ErrorMatches, "internal error: cannot use with filesystem structures")
+	c.Check(found, Equals, "")
+	c.Check(offs, Equals, gadget.Size(0))
+}
+
+func (d *deviceSuite) TestDeviceFindFallbackBadEvalSymlinks(c *C) {
+	if os.Geteuid() == 0 {
```

Comment:
Can we instead have `var evalSymlinks = filepath.EvalSymlinks` and mock it in the test to inject an error, so this restriction can be avoided?

Reply:
added